LEFT JOIN null-extension on inner-side VARCHAR can panic with UNION ALL and ORDER BY

## Bug Report

### Summary

TiDB can panic when a `LEFT JOIN` produces null-extended rows for an inner-side string column, and the result is later passed through `UNION ALL` plus `ORDER BY`.

This repro does not require generated columns. A plain inner-side `VARCHAR` column is enough.

On my side, the server panics in result encoding with:

```text
runtime error: index out of range [2] with length 2
```

### 1. Minimal reproduce step

```sql
DROP DATABASE IF EXISTS codex_plain_probe;
CREATE DATABASE codex_plain_probe;
USE codex_plain_probe;

CREATE TABLE t_outer (
  id INT NOT NULL
);

CREATE TABLE t_inner (
  id INT NOT NULL,
  d VARCHAR(64) NOT NULL
);

INSERT INTO t_outer VALUES (1), (3);
INSERT INTO t_inner VALUES (2, 'alpha');

SELECT d
FROM (
  (
    SELECT t_inner.d
    FROM t_outer
    LEFT JOIN t_inner ON t_outer.id = t_inner.id
    WHERE t_inner.id IS NULL
  )
  UNION ALL
  SELECT 'z'
) y
ORDER BY d;
```

### 2. What did you expect to see?

The query should return 3 rows without panic:

```text
NULL
NULL
z
```

### 3. What did you see instead?

```text
ERROR 1105 (HY000): runtime error: index out of range [2] with length 2
```

### 4. What is your TiDB version?

```text
Release Version: v9.0.0
Edition: Community
Git Commit Hash: 4ae7a12ae7884061a9b095584fd22bc3eab23112
Git Branch: heads/refs/tags/v9.0.0
UTC Build Time: 2026-04-25 02:22:25
GoVersion: go1.25.9
Race Enabled: false
Check Table Before Drop: false
Store: tikv
Kernel Type: Classic
```

### TiDB log / panic stack

```text
[2026/04/29 07:17:09.151 +00:00] [ERROR] [conn.go:1080] ["connection running loop panic"] [conn=2843837430] [session_alias=] [lastSQL="select d from ((select t_inner.d from t_outer left join t_inner on t_outer.id = t_inner.id where t_inner.id is null) union all select 'z') y order by d"] [err="runtime error: index out of range [2] with length 2"] [stack="github.com/pingcap/tidb/pkg/server.(*clientConn).Run.func1
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn.go:1083
runtime.gopanic
	/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.9.linux-amd64/src/runtime/panic.go:783
github.com/pingcap/tidb/pkg/server.(*clientConn).writeResultSet.func1
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn.go:2357
runtime.gopanic
	/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.9.linux-amd64/src/runtime/panic.go:783
runtime.goPanicIndex
	/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.9.linux-amd64/src/runtime/panic.go:115
github.com/pingcap/tidb/pkg/util/chunk.(*Column).GetBytes
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/chunk/column.go:737
github.com/pingcap/tidb/pkg/util/chunk.Row.GetBytes
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/chunk/row.go:81
github.com/pingcap/tidb/pkg/server/internal/column.DumpTextRow
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/internal/column/column.go:200
github.com/pingcap/tidb/pkg/server.(*clientConn).writeChunks
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn.go:2491
github.com/pingcap/tidb/pkg/server.(*clientConn).writeResultSet
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn.go:2376
github.com/pingcap/tidb/pkg/server.(*clientConn).handleStmt
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn.go:2161
github.com/pingcap/tidb/pkg/server.(*clientConn).handleQuery
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn.go:1870
github.com/pingcap/tidb/pkg/server.(*clientConn).dispatch
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn.go:1460
github.com/pingcap/tidb/pkg/server.(*clientConn).Run
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/conn.go:1196
github.com/pingcap/tidb/pkg/server.(*Server).onConn
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/server/server.go:791"]
```

--- a/pkg/planner/core/casetest/rule/rule_outer_to_semi_join_test.go
+++ b/pkg/planner/core/casetest/rule/rule_outer_to_semi_join_test.go
@@ -62,6 +62,29 @@ func TestOuterToSemiJoin(tt *testing.T) {
 			tk.MustQuery("EXPLAIN FORMAT='plan_tree' " + sql).Check(testkit.Rows(output[i].Plan...))
 			tk.MustQuery(sql).Check(testkit.Rows(output[i].Result...))
 		}
+
+		// issue:68112
+		tk.MustExec("drop table if exists t_outer, t_inner")
+		tk.MustExec("create table t_outer (id int not null)")
+		tk.MustExec(`create table t_inner (
+			id int not null,
+			d varchar(64) not null
+		)`)
+		tk.MustExec("insert into t_outer values (1), (3)")
+		tk.MustExec("insert into t_inner values (2, 'alpha')")
+		query := `select d
+			from (
+				(
+					select t_inner.d
+					from t_outer
+					left join t_inner on t_outer.id = t_inner.id
+					where t_inner.id is null
+				)
+				union all
+				select 'z'
+			) y
+			order by d`
+		tk.MustQuery(query).Check(testkit.Rows("<nil>", "<nil>", "z"))
 	})
 }
 

--- a/pkg/planner/core/rule/rule_outer_join_to_semi_join.go
+++ b/pkg/planner/core/rule/rule_outer_join_to_semi_join.go
@@ -385,6 +385,9 @@ func generateProjectForConvertAntiJoin(p *logicalop.LogicalJoin, innerSchSet *in
 	projExprs := make([]expression.Expression, 0, len(parentNodeSchema.Columns))
 	for _, c := range parentNodeSchema.Columns {
 		if innerSchSet.Has(int(c.UniqueID)) {
+			// Keep null-extended inner columns aligned with the parent schema type.
+			// NewNull() defaults to TINYINT, while the original column can have
+			// another type and must become nullable after the outer join rewrite.
 			retType := c.RetType.Clone()
 			retType.DelFlag(mysql.NotNullFlag)
 			projExprs = append(projExprs, expression.NewNullWithFieldType(retType))

--- a/pkg/planner/core/rule/rule_outer_join_to_semi_join.go
+++ b/pkg/planner/core/rule/rule_outer_join_to_semi_join.go
@@ -385,7 +385,9 @@ func generateProjectForConvertAntiJoin(p *logicalop.LogicalJoin, innerSchSet *in
 	projExprs := make([]expression.Expression, 0, len(parentNodeSchema.Columns))
 	for _, c := range parentNodeSchema.Columns {
 		if innerSchSet.Has(int(c.UniqueID)) {
-			projExprs = append(projExprs, expression.NewNull())
+			retType := c.RetType.Clone()
+			retType.DelFlag(mysql.NotNullFlag)
+			projExprs = append(projExprs, expression.NewNullWithFieldType(retType))
 		} else {
 			projExprs = append(projExprs, c.Clone())
 		}

